### ISSUE TITLE
"pudata was nullptr"

### ISSUE THREAD

**Issue by laxeot** [ThePhD/sol2#101]
**EDIT: ** I literally forgot to explain the error. Scroll to the bottom of this post for it.

Hi,
I have a class in my C++ project called "LuaController" with a little test function called "spit".

**LuaController.h** (with functions unnecessary to the problem removed)
```
public:
	LuaController(std::string n);
	void setName(std::string s);
	std::string getName();
	void spit(std::string s);
private:
	std::string name;
```

**LuaController.cpp** (same deal with the functions)
```
LuaController::LuaController(std::string n)
{
	name = n;
}
void LuaController::setName(std::string s)
{
	name = s;
}
std::string LuaController::getName()
{
	return name;
}
void LuaController::spit(std::string s)
{
	std::cout << s << "\n";
}
```

Then, I create a usertype for the script. Everything works, EXCEPT the spit function.
```
lua.new_usertype<LuaController>("Controller",
	sol::constructors<sol::types<std::string>>(),
	"spit", &LuaController::spit,
	"name", sol::property(&LuaController::getName, &LuaController::setName)
);
```

Spit works TOTALLY fine if it doesn't have any arguments (i.e., just printing something with no user input). Does sol2 not allow arguments to be passed to functions?

**page1.lua**
```
controller = Controller.new("testController")

print("The name of the controller is " .. controller.name)
controller.name = "luaController"
print("The name of the controller is " .. controller.name)

controller.spit("Hello, world!")
```

Basically, what happens is when `controller.spit("Hello, world!")` is called, I get a read access violation with the message "pudata was nullptr". Any idea what I'm doing wrong?

**Comment by ThePhD** [comment]
It should be `controller:spit`. Lua doesn't pass the `self` argument (`controller` in this case) unless you use that special syntax.

I have not yet made a release for my current roll of changes. One of them is going to include an explicit warning for this in debug mode turned on by default.

**Comment by ThePhD** [comment]
@LittleKu You should open a new issue for this, rather than comment in a closed one.

For me it seems like your problem is that you're returning a `(TCHAR*)`. A `TCHAR*` pointer may not be a recognizable string type. You cannot print that.

I also don't know what `SOUI::S_CA2T` does, or what `SStringT` is. These seem like poor wrappers around Windows-based conversion utilities, but I am not sure and I cannot duplicate your problem in a test program.

Lua's strings are canonically stored and referenced as utf8. If you want to convert it to some other representation, you can, but `print()` and other functions are liable not to understand it or display it well on your console/screen.

**Comment by LittleKu** [comment]
ok, maybe a demo wiil be better for you that solve my problem, will open a new issue, thx ;)
